std/utf.d

```diff
@@ -3348,85 +3349,79 @@ enum dchar replacementDchar = '\uFFFD';
  *      $(REF byGrapheme, std,uni).
  */
 auto byCodeUnit(R)(R r)
-if (isAutodecodableString!R)
+if (isAutodecodableString!R ||
+    isInputRange!R && isSomeChar!(ElementEncodingType!R) ||
+    (is(R == enum) && is(R : const dchar[])))
 {
-    static struct ByCodeUnitImpl
+    static if (isNarrowString!R ||
+               // This would be cleaner if we had a way to check whether a type
+               // was a range without any implicit conversions.
+               (isAutodecodableString!R && !__traits(hasMember, R, "empty") &&
+                !__traits(hasMember, R, "front") && !__traits(hasMember, R, "popFront")))
     {
-    pure nothrow @nogc:
-
-        @property bool empty() const         { return r.length == 0; }
-        @property auto ref front() inout     { return r[0]; }
-        void popFront()                      { r = r[1 .. $]; }
-        auto ref opIndex(size_t index) inout { return r[index]; }
-
-        @property auto ref back() inout
+        static struct ByCodeUnitImpl
         {
-            return r[$ - 1];
-        }
+        @safe pure nothrow @nogc:
 
-        void popBack()
-        {
-            r = r[0 .. $-1];
-        }
+            @property bool empty() const     { return str.length == 0; }
+            @property auto ref front() inout { return str[0]; }
+            void popFront()                  { str = str[1 .. $]; }
 
-        static if (!isAggregateType!R)
-        {
-            auto opSlice(size_t lower, size_t upper)
-            {
-                return ByCodeUnitImpl(r[lower .. upper]);
-            }
-        }
+            @property auto save() { return ByCodeUnitImpl(str.save); }
 
-        @property size_t length() const
-        {
-            return r.length;
-        }
-        alias opDollar = length;
+            @property auto ref back() inout { return str[$ - 1]; }
+            void popBack()                  { str = str[0 .. $-1]; }
 
-        static if (!isAggregateType!R)
-        {
-            @property auto save()
-            {
-                return ByCodeUnitImpl(r.save);
-            }
-        }
+            auto ref opIndex(size_t index) inout     { return str[index]; }
+            auto opSlice(size_t lower, size_t upper) { return ByCodeUnitImpl(str[lower..upper]); }
 
-      private:
-        R r;
-    }
+            @property size_t length() const { return str.length; }
+            alias opDollar = length;
 
-    static assert(isAggregateType!R || isRandomAccessRange!ByCodeUnitImpl);
+          private:
+            StringTypeOf!R str;
+        }
 
-    return ByCodeUnitImpl(r);
-}
+        static assert(isRandomAccessRange!ByCodeUnitImpl);
 
-/// Ditto
-auto ref byCodeUnit(R)(R r)
-if (!isAutodecodableString!R && isInputRange!R && isSomeChar!(ElementEncodingType!R))
-{
-    // byCodeUnit for ranges and dchar[] is a no-op
-    return r;
+        return ByCodeUnitImpl(r);
+    }
+    else static if (is(R == enum) || is(R : const dchar[]))
```

Comment:
LOL. Ah, I see now. Yeah, very close, and I missed the difference upon review. And now you're making me think. :)

Well, the `&&` is definitely required in the template constraint, otherwise any old enum would pass, which would obviously be bad. We only want enums that are strings to pass. `isAutodecodableString` does that already for enums of arrays of `char` and `wchar`. The `is(R == enum) && is(R : const dchar[])` bit is supposed to then cover enums of arrays of `dchar`. However, `isAutodecodableString` also lets user-defined types that implicitly convert to narrow strings through (but not static arrays). So, it would appear that I made an oversight with regards to types that implicitly convert to arrays of `dchar`. What that really should be is `is(R : const dchar[]) && !isStaticArray!R`, because that would be the `dchar` equivalent of what `isAutodecodableString` does. The `isInputRange!R && isSomeChar!(ElementEncodingType!R)` lets any ranges of any character type through, and that should be fine.

As for the static if-else chain, the first branch needs to be for narrow strings and anything that implicitly converts to narrow strings - excepting the bizarre corner case where a user-defined type is already a range but also implicitly converts to a narrow string. For this branch, the implicit conversion is enforced (which is a no-op for actual narrow strings), and then we get the wrapper range that avoids auto-decoding. I see nothing wrong with that `static if` condition. Either it's a narrow string, or it's an enum or user-defined type which implicitly converts to a narrow string and is not itself a range. 

That leaves two branches. One of them must cover anything that implicitly converts to an array of `dchar` - but is not itself a range - and the other must cover any ranges of characters. Actual arrays of `dchar` that aren't enums could actually go in either branch. As it stands, the fact that `||` is used makes arrays of `dchar` take the first branch. It also makes it so that ranges that implicitly convert to an array of `dchar` get converted (which we don't want). And the `enum` actually ends up being pointless. If it were `&&` like in the template constraint, then _only_ enums would get through, and actually, since only enums that implicitly converted to an array of `dchar` could get to this condition (since any other enum type would not be a range and thus would have failed the template constraint, and all of the enums that implicitly converted to narrow strings went with the first branch), simply checking `is(R == enum)` would be the same as changing it to `&&`. But if we changed it to `&&`, then user-defined types that were not ranges in and of themselves but implicitly converted to an array of `dchar` would simply be returned rather than converted, which would be bad. So, neither `||` nor `&&` is correct. I'll have to adjust it so that it does the conversion for user-defined types only if they're not already ranges while still doing the conversion for those that aren't already ranges.

So, if I've thought this through correctly, ultimately, both were conditions involving `is(R : const dchar[])` were wrong. Man, I hate this maze of traits involving strings, and implicit conversions are hell for generice code. :|

I'll fix this up later today. Thanks for catching that.